### DISCUSSION TITLE
Can DefaultThreadContextMap be modified to avoid creating a new instance with each update?

### DISCUSSION THREAD

**Discussion by wang-qijia** [apache/logging-log4j2#3032] [Q&A]
If the client calls MDC.put() 10 times, the current implementation creates 10 maps, resulting in frequent memory allocations. Is it possible to modify this so that put does not create a new map, and enhance getImmutableMapOrNull to return a decorated immutable version of the original map?

**Comment by ppkarwasz** [comment]
Hi @wang-qijia,

The `DefaultThreadContextMap` is (and always has been) a copy-on-write structure, which is useful if you need to move it from thread to thread.

If you want to **limit** the number of allocations, you can use methods such as [`ThreadContext.putAll()`](https://logging.apache.org/log4j/2.x/javadoc/log4j-api/org/apache/logging/log4j/ThreadContext.html#putAll(java.util.Map)).

If you want to **remove** the allocation entirely and you are **not using** a Servlet Container, you can switch to a garbage-free implementation as described in [garbage-free logging](https://logging.apache.org/log4j/2.x/manual/garbagefree.html) by setting the [`log4j2.threadContextMap` Log4j property](https://logging.apache.org/log4j/2.x/manual/garbagefree.html#log4j2.threadContextMap) to `GarbageFree` (or `org.apache.logging.log4j.spi.GarbageFreeSortedArrayThreadContextMap` if you are using Log4j 2.23.1 or an earlier version).

**Note**: The garbage-free `ThreadContextMap` is not always the best solution, that is why it is not the default one. What kind of application are you developing? Is being garbage-free important for you application? Garbage-free implementations always come at the cost of an increased memory usage.

**Reply by wang-qijia** [reply]
Hi @ppkarwasz 
Thank you for your participation in the discussion.

The use case is to populate the MDC context into logs, and modifying the specific scenario using ThreadContext.putAll() would have a high cost. The described scenario can be achieved through other means, and as I understand it, it does not rely on a copy-on-write structure.

**Reply by ppkarwasz** [reply]
Yes, the easiest solution for you is to switch to a garbage-free `ThreadContextMap`.

`ThreadContextMap` is one of the most update interfaces in Log4j API. Adapting `DefaultThreadContextMap` in such a way that it:

- does not create memory leaks in Servlet environments (i.e. it can only store JRE classes in `ThreadLocal`s),
- it is mutable,
- yet, it does not require a **copy operation**, when it is transfered to an asynchronous thread for logging.

is IMHO a highly non-trivial operation.

**Comment by vy** [comment]
@wang-qijia, I tend to agree with you. I think the default Thread Context implementation can (and should?) do better – see [my comments here](https://github.com/apache/logging-log4j2/pull/2330#issuecomment-2388334460). I intend to submit a PR to improve the situation.

**Reply by wang-qijia** [reply]
Hi @vy 
Thank you for your participation in the discussion.

Has there been any progress on this? Do you still need my help?

**Reply by vy** [reply]
@wang-qijia, thanks for asking, not needed at this point.
